### DISCUSSION TITLE
Questions About DriveBase on Spike Prime

### DISCUSSION THREAD

**Discussion by hashen0608** [pybricks/support#1649] [Q&A] [answered]
Hello,

I am about to guide my students into using Pybricks, but I encountered an issue with moving straight forward and backward. Just like in the video, my robot veers off course when using the gyroscope to move forward and backward. However, there doesn't seem to be any problem with the program itself.

I would like to ask everyone for any thoughts and suggestions. Thank you all.

I have attached the video and program for everyone's reference and brainstorming.

![image](https://github.com/pybricks/support/assets/167075380/ae3316be-5afc-4509-b20b-d7ba868f9b95)

https://github.com/pybricks/support/assets/167075380/8d14d61b-4933-4897-9098-ff8127639111

**Comment by laurensvalk** [comment, marked as answer]
Thanks for sharing this. At first glance it looked unusual, but I think there is a straightforward explanation.

You've configured the drive speed at 620 mm/s. With 56 mm wheels, that amounts to about 1268 degrees per second, which is far more than the LEGO Spike motors can do (about 1000 deg/s). With some load, they would go even slower.

At that speed, there is no more room for automatic speed/direction correction, since it is already going at maximum speed.

This is why we're seeing it happen in the middle of each run in your video, which is where it hits top speed. As it slows down, it has some breathing room for corrections again, so we see it recover.

----------------

I replicated your program but added an extra parallel loop to print the motor speeds, which makes it possible to see it without doing the math. As you can see below, it peaks at about 1000 deg/s, which is about the limit for these motors. And this is if I lift it up. Under a heavy robot, the top speed is usually lower.

![image](https://github.com/pybricks/support/assets/12326241/e92fba2f-67d5-47a9-8bc1-9ed1d7d92097)

When I change your program to run at 300 mm/s (30 cm/s is still pretty fast!), it does go straight.

At 400 mm/s, which is closer to the theoretical limit (about 820 deg/s), it also still goes straight, but I would probably avoid running on the limit.

You could go a lot faster with bigger wheels, but that's a design trade off against other aspects of your robot.

**Reply by nirbhuyjillu** [reply]
"You've configured the drive speed at 620 mm/s. With 56 mm wheels, that amounts to about 1544 degrees per second, which is far more than the LEGO Spike motors can do (about 1000 deg/s)." --- could you please explain the math?

**Reply by laurensvalk** [reply]
Looks like there was a typo in my original answer. I was working with a diameter of 46mm instead of 56mm, thanks for catching it! The conclusion still stands, though.

-------------------

Here's the slightly longer version that is perhaps more intuitive to follow. You can add a pointer or marker to the wheel to visualize it.

- A speed of 620 mm/s is the speed of the robot. Or viewed on board the robot, it's how fast the floor is passing by.
- So if the wheels don't slip, the ground contact point is always moving at 620 mm/s.
- The wheel is rigid, so all outer points on the wheel are going at the same speed.
- The wheel diameter is `56 mm` so the circumference is `π * 56 mm = 175.93 mm`.
- So for one point on the wheel to make one full circle, we can use the standard formula for distance and speed: `distance = speed * time`. Or equivalently the time for one full circle: `time = distance / speed`.
- So one full wheel revolution takes `time = distance / speed = 175.93 / 620` = 0.2837` seconds.
- One full rotation is 360 degrees. So it is doing 360 degrees per `0.2837` second. So it is rotating at `360 / 0.2837 = 1268` degrees per second.

-----------

The short version is to multiply the configured speed by `(360 / π / diameter)`. That gives you `620 * 360 / π / 56 = 1268 deg/s` directly.

**Reply by nirbhuyjillu** [reply]
Spike prime efficiency is 135 to 185 RPM.

185 * 360 / 60 deg/s
That ranges from 810 to 1110 degrees/s

https://le-www-live-s.legocdn.com/sc/media/files/support/spike-prime/techspecs_technicmediumangularmotor-19684ffc443792280359ef217512a1d1.pdf

LEGO® Education SPIKE™ Prime Technical Specifications

**Reply by laurensvalk** [reply]
Yes, under no load it can go that fast, but most applications have some load.

**Comment by neptunus0001** [comment]
This explanation has been mega helpful to me! Thank you all!

It would be of great value, if there was an explanation for turn speed also on the page. Can anyone help with that?

`straight_speed` in `mm/s` Is explained in details.
`straight_acceleration` in `mm/s²`  `turn_rate` in `deg/s`  `turn_acceleration` in `deg/s²` 
Are not really explained in detail therefor it would be great if there was an detailed explanation on how to set this up for a robot. Or at 
least some pointers in the right direction would be awesome.


Below I have included two tables that I made for my own project. This may help others.

Technic™ Medium Angular Motor | [link](https://le-www-live-s.legocdn.com/sc/media/files/support/spike-prime/techspecs_technicmediumangularmotor-19684ffc443792280359ef217512a1d1.pdf) |   |  
-- | -- | -- | --
RPM | 185 | 1110 | degrees/s |  
+/- | 15 | % |   |   |  
min | 157 | 944 | degrees/s |  
max | 209 | 1252 | degrees/s |  


Technic™  Large Angular Motor | [link](https://le-www-live-s.legocdn.com/sc/media/files/support/spike-prime/techspecs_techniclargeangularmotor-1b79e2f4fbb292aaf40c97fec0c31fff.pdf)  |   |  
-- | -- | -- | --
RPM | 175 | 1050 | degrees/s |  
+/- | 15 | % |   |   |  
min | 149 | 893 | degrees/s |  
max | 197 | 1184 | degrees/s |  

 drive speed | 400 | 300 | 620 | 300
-- | -- | -- | -- | --
 wheel diameter | 62,5 | 62,5 | 56 | 56
 deg/s | 733 | 550 | 1269 | 614

**Reply by zoligyenge** [reply]
Thanks. Do you have any data on max angular accelerations of the motors? I know it depends on load, but some rough estimates
